src/utility/HCISharedMemTransport.cpp

```diff
@@ -0,0 +1,746 @@
+
+    WirelessFwInfo_t wireless_info_instance;
+    WirelessFwInfo_t *p_wireless_info = &wireless_info_instance;
+    SHCI_GetWirelessFwInfo(p_wireless_info);
+#if defined(PRINT_IPCC_INFO)
+    printf("WB copro FW version = %d.%d.%d\r\n", p_wireless_info->VersionMajor, p_wireless_info->VersionMinor, p_wireless_info->VersionSub);
+#endif /*(PRINT_IPCC_INFO)*/
+
+    /* Now start BLE service on firmware side, using Vendor specific
+     * command on the System Channel
+     */
+    status = stm32wb_start_ble();
+
+    /* Once reset complete event is received we will need
+     * to send a few more commands:
+     * set bd addr with bt_ipm_set_addr();
+     * during the HCI rest command */
+  }
+  /* IPM Channel is now open */
+#if defined(PRINT_IPCC_INFO)
+  printf("IPM Channel Open Completed\r\n");
+#endif /*(PRINT_IPCC_INFO)*/
+
+  return status;
+}
+
+void HCISharedMemTransportClass::end()
+{
+  /* M0 sub-system is already on (sys_event) */
+  sys_resp = false;
+  acl_data_on = false;
+  data_overflow = false;
+
+  /* the HCI RESET command ready to be processed again */
+  phase_bd_addr = false;
+  phase_tx_power = false;
+  phase_reset = false;
+  phase_running = false;
+}
+
+void HCISharedMemTransportClass::wait(unsigned long timeout)
+{
+  for (unsigned long start = millis(); (millis() - start) < timeout;) {
+    if (available()) {
+      break;
+    }
+  }
+}
+
+int HCISharedMemTransportClass::available()
+{
+  /* assuming the reset is already achieved,
+   * the LL-only mode is already configured. */
+
+  if (_read_index != _write_index) {
+    return 1;
+  } else if (data_overflow) {
+    data_overflow = false;
+    if (_read_index != _write_index) {
+      return 1;
+    }
+  }
+
+  return 0;
+}
+
+int HCISharedMemTransportClass::peek()
+{
+  int peek_val = -1;
+  __disable_irq();
+  if (_read_index != _write_index) {
+    peek_val = _rxbuff[_read_index];
+  }
+  __enable_irq();
+  return peek_val;
+}
+
+int HCISharedMemTransportClass::read()
+{
+  int read_val = -1;
+  __disable_irq();
+  if (_read_index != _write_index) {
+    read_val = _rxbuff[_read_index];
+    _read_index++;
+    if (_read_index == _write_index) {
+      /* Reset buffer index */
+      _read_index = 0;
+      _write_index = 0;
+    }
+  }
+  __enable_irq();
+  return read_val;
+}
+
+size_t HCISharedMemTransportClass::write(const uint8_t *data, size_t length)
+{
+  const uint8_t *msg_data;
+  msg_data = &data[1];
+
+  /* capture the HCI reset send command opcode = 0x0C03
+   * After HCI reset event complete in the evt_received(),
+   * the bd_addr and tx_power must be sent
+   * before the phase_running begins.
+   */
+  if (phase_running) {
+    return mbox_write(data[0], (length - 1), msg_data);;
+  }
+  if ((data[1] == 0x03) && (data[2] == 0x0C)) {
+    phase_reset = false;
+
+    mbox_write(data[0], (length - 1), msg_data);
+
+    /* capture event after HCI_RESET */
+    while (!phase_reset);
+
+    /* set the bd add */
+    bt_ipm_set_addr();
```

Comment:
this function returns an error code, probably it should be checked.

Reply:
ok, in this case of error, will return 'no data written'